File: frontend/src/api/retrieval.ts
/** Retrieval endpoints - EchoNotes Feature 4. */

import type { Note } from "@/types";

export interface RetrievedNote {
  note: Note;
  subjectName: string;
  topicName: string;
  score: number;
}

export interface QueryResult {
  intent: string;
  spoken: string;
  answer: string | null;
  results: RetrievedNote[];
  sources: string[];
  navigateTo: string | null;
}

export function askByText(_utterance: string): Promise<QueryResult> {
  throw new Error("not implemented");
}

export function askByVoice(_audio: Blob): Promise<QueryResult> {
  throw new Error("not implemented");
}

File: backend/app/schemas/retrieval.py
"""Retrieval schemas - EchoNotes Feature 4."""

from __future__ import annotations

from pydantic import BaseModel

from app.schemas.note import NoteOut


class VoiceQuery(BaseModel):
    """Either an already-transcribed utterance, or audio uploaded to be transcribed first."""

    utterance: str
    subject_id: str | None = None
    top_k: int = 5


class RetrievedNote(BaseModel):
    note: NoteOut
    subject_name: str
    topic_name: str
    score: float


class QueryResult(BaseModel):
    intent: str
    spoken: str  # the complete response to speak
    answer: str | None = None
    results: list[RetrievedNote] = []
    sources: list[str] = []
    navigate_to: str | None = None  # element id for the client to focus
